static/app/components/commandPalette/context.tsx

```diff
@@ -0,0 +1,84 @@
+import {createContext, useCallback, useContext, useMemo, useState} from 'react';
+
+import type {CommandPaletteAction} from './types';
+
+type CommandPaletteProviderProps = {children: React.ReactNode};
+
+type CommandPaletteStore = {
+  actions: CommandPaletteAction[];
+};
+
+type CommandPaletteConfig = {
+  registerActions: (actions: CommandPaletteAction[]) => void;
+  unregisterActions: (keys: string[]) => void;
+};
+
+const CommandPaletteConfigContext = createContext<CommandPaletteConfig | null>(null);
+const CommandPaletteStoreContext = createContext<CommandPaletteStore | null>(null);
+
+export function useCommandPaletteConfiguration(): CommandPaletteConfig {
+  const ctx = useContext(CommandPaletteConfigContext);
+  if (ctx === null) {
+    throw new Error('Must be wrapped in CommandPaletteProvider');
+  }
+  return ctx;
+}
+
+export function useCommandPaletteStore(): CommandPaletteStore {
+  const ctx = useContext(CommandPaletteStoreContext);
+  if (ctx === null) {
+    throw new Error('Must be wrapped in CommandPaletteProvider');
+  }
+  return ctx;
+}
+
+export function CommandPaletteProvider({children}: CommandPaletteProviderProps) {
+  const [actions, setActions] = useState<CommandPaletteAction[]>([]);
```

Comment:
The action and usage pattern here would be more suitable for a reducer than useState. That way all that you need to pass to the reducers is just the dispatch function and the state, and you won't need to deal with any memo problems